core/domain/stats_jobs_one_off.py

```diff
@@ -28,14 +34,248 @@
+                the old_stats_model up to the next version.
+
+        Returns:
+            dict. An dict representation of an ExplorationStatsModel
+                with updated state_stats_mapping and version.
+        """
+        # Handling state additions, deletions and renames.
+        for change_dict in change_list:
+            if change_dict['cmd'] == exp_domain.CMD_ADD_STATE:
+
+                old_stats_model['state_stats_mapping'][change_dict[
+                    'state_name']] = (stats_domain.StateStats
+                                      .create_default())
+            elif change_dict['cmd'] == exp_domain.CMD_DELETE_STATE:
+                old_stats_model['state_stats_mapping'].pop(change_dict[
+                    'state_name'])
+            elif change_dict['cmd'] == exp_domain.CMD_RENAME_STATE:
+                old_stats_model['state_stats_mapping'][
+                    change_dict['new_state_name']] = (
+                        old_stats_model['state_stats_mapping'].pop(
+                            change_dict['old_state_name']))
+        old_stats_model['exp_version'] += 1
+
+        return old_stats_model
+
+    @classmethod
+    def _recompute_statistics(
+            cls, exp_id, exp_version, events, old_stats_model):
+        """Creates a dict representing the attributes of an
+        ExplorationStatsModel instance.
+
+        Args:
+            exp_id: str. The id of the exploration to recompute statistics for.
+            exp_version. The version of the exploration.
+            events: list(dict). A list of event data dicts from the datastore.
+            old_stats_model: A dict representing the ExplorationStatsModel
+                instance.
+
+        Returns:
+            dict. A dict representing the attributes of an ExplorationStatsModel
+                instance.
+        """
+        state_hit_ids, solution_hit_ids = set(), set()
+
+        exp_stats_dict = {}
+        for event in events:
+            state_stats = (old_stats_model['state_stats_mapping'][
+                event['state_name']])
+            if event['event_type'] == feconf.EVENT_TYPE_ANSWER_SUBMITTED:
+                state_stats['total_answers_count_v2'] += 1
+                if event['is_feedback_useful']:
+                    state_stats['useful_feedback_count_v2'] += 1
+            elif event['event_type'] == feconf.EVENT_TYPE_STATE_HIT:
+                state_stats['total_hit_count_v2'] += 1
+                if event['session_id'] not in state_hit_ids:
+                    state_stats['first_hit_count_v2'] += 1
+                    state_hit_ids.add(event['session_id'])
+            elif event['event_type'] == feconf.EVENT_TYPE_SOLUTION_HIT:
+                if event['session_id'] not in solution_hit_ids:
+                    state_stats['num_times_solution_viewed_v2'] += 1
+                    solution_hit_ids.add(event['session_id'])
+            elif event['event_type'] == (
```

Comment:
Suggest moving these two to the top, to distinguish them from the ones that actually need state_stats.

Reply:
Done.